//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.queryserver";
option java_outer_classname = "QueryServer";
option objc_class_prefix = "LVQS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package queryserver;

//---------------------------------------------------------------------
// The QueryServer service definition.
//---------------------------------------------------------------------
service QueryServer {
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  rpc Query (QueryRequest) returns (QueryResponse) {}
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}

  rpc TestDataTypes(TestDataTypesParameters) returns (TestDataTypesParameters) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message DataTypeTest {
  bool boolValaue = 1;
  int32 intValue = 2;
  float floatValue = 3;
  double doubleValue = 4;
  // string stringValue = 5;
  // bytes bytesValue = 6;
  // repeated bool repeatedBoolValaue = 7;
  // repeated int32 repeatedIntValue = 8;
  // repeated float repeatedFloatValue = 9;
  // repeated double repeatedDoubleValue = 10;
  // repeated string repeatedStringValue = 11;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message TestDataTypesParameters {
  string root = 1;
  repeated string RepeatedString = 2;
  repeated int32 RepeatedIntValue = 3;
  int32 rootInt32 = 4;
  bool rootBool = 5;
  repeated double RepeatedDoubleValue = 6;
  double rootDouble = 7;
  repeated bool RepeatedBoolValue = 8;
  float rootFloat = 9;
  repeated float RepeatedFloatValue = 10;
  DataTypeTest nested = 11;
  repeated DataTypeTest repeatedNested = 12;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int32 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryResponse {
  string message = 1;
  int32 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int32 serverId = 2;
  int32 status = 3;
}
